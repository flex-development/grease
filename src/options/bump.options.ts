/**
 * @file Options - BumpOptions
 * @module grease/options/BumpOptions
 */

import { IsManifestId, IsReleaseVersion } from '#src/decorators'
import type { ReleaseVersion } from '#src/types'
import type { ModuleId } from '@flex-development/mlly'
import { DOT, defaults, type Assign } from '@flex-development/tutils'
import { IsBoolean, IsIn, IsString } from 'class-validator'

/**
 * User version bump options.
 */
type BumpOptionsDTO = Assign<Partial<BumpOptions>, {
  /**
   * Release type or version.
   */
  release: BumpOptions['release']
}>

/**
 * Version bump options.
 *
 * @class
 */
class BumpOptions {
  /**
   * Module id of package manifest or directory.
   *
   * @default DOT
   *
   * @public
   * @instance
   * @member {ModuleId} manifest
   */
  @IsManifestId()
  public manifest: ModuleId

  /**
   * Prerelease identifier.
   *
   * @see https://docs.npmjs.com/cli/commands/npm-dist-tag
   *
   * @default 'alpha'
   *
   * @public
   * @instance
   * @member {string} preid
   */
  @IsString()
  public preid: string

  /**
   * Base to be used for prerelease identifier.
   *
   * @default 1
   *
   * @public
   * @instance
   * @member {0 | 1} prestart
   */
  @IsIn([0, 1])
  public prestart: 0 | 1

  /**
   * Release type or version.
   *
   * @public
   * @instance
   * @member {ReleaseVersion} release
   */
  @IsReleaseVersion()
  public release: ReleaseVersion

  /**
   * Disable logs.
   *
   * @default false
   *
   * @public
   * @instance
   * @member {boolean} silent
   */
  @IsBoolean()
  public silent: boolean

  /**
   * Write version bump to package manifest.
   *
   * @default false
   *
   * @public
   * @instance
   * @member {boolean} write
   */
  @IsBoolean()
  public write: boolean

  /**
   * Create a new options object.
   *
   * @see {@linkcode BumpOptionsDTO}
   *
   * @param {BumpOptionsDTO} opts - User version bump options
   */
  constructor(opts: BumpOptionsDTO) {
    const {
      manifest,
      preid,
      prestart,
      release,
      silent,
      write
    } = defaults(opts, {
      manifest: DOT,
      preid: 'alpha',
      prestart: 1 as const,
      silent: false,
      write: false
    })

    this.manifest = manifest
    this.preid = preid
    this.prestart = prestart
    this.release = release
    this.silent = silent
    this.write = write
  }
}

export { BumpOptions as default, type BumpOptionsDTO }
