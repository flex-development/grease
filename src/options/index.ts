/**
 * @file Entry Point - Options
 * @module grease/options
 */

export { default as GlobalOptions } from './global.options'
