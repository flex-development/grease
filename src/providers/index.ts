/**
 * @file Entry Point - Providers
 * @module grease/providers
 */

export { default as ValidationService } from './validation.service'
