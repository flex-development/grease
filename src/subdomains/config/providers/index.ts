/**
 * @file Entry Point - Providers
 * @module grease/config/providers
 */

export { default as ConfigService } from './config.service'
