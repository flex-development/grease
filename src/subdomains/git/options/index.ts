/**
 * @file Entry Point - Options
 * @module grease/git/options
 */

export { default as CommitGrammarOptions } from './commit-grammar.options'
export { default as GitCommitOptions } from './git-commit.options'
export { default as GitTagOptions } from './git-tag.options'
export { default as GitOptions } from './git.options'
