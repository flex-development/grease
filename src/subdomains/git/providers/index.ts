/**
 * @file Entry Point - Providers
 * @module grease/git/providers
 */

export { default as GitService } from './git.service'
