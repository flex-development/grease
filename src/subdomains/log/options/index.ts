/**
 * @file Entry Point - Options
 * @module grease/log/options
 */

export { default as LoggerOptions } from './logger.options'
