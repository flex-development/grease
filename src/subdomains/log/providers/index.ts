/**
 * @file Entry Point - Providers
 * @module grease/log/providers
 */

export { default as LoggerService } from './logger.service'
