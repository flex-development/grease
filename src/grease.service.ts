/**
 * @file GreaseService
 * @module grease/GreaseService
 */

import {
  BumpEvent,
  BumpOperation,
  BumpQuery,
  RecommendedBump,
  type BumpOperationDTO
} from '#src/bump'
import {
  ChangelogEvent,
  ChangelogOperation,
  type ChangelogOperationDTO,
  type ChangelogStream
} from '#src/changelog'
import {
  ConfigService,
  type GreaseConfig,
  type IGreaseConfig
} from '#src/config'
import {
  TagOperation,
  TagQuery,
  type Commit,
  type TagOperationDTO
} from '#src/git'
import type { PackageManifest } from '#src/models'
import { LoggerService } from '#src/providers'
import { isString, type EmptyObject } from '@flex-development/tutils'
import { Injectable } from '@nestjs/common'
import { CommandBus, EventBus, QueryBus } from '@nestjs/cqrs'
import { GlobalOptions } from './options'

/**
 * Grease runner service.
 *
 * @class
 */
@Injectable()
class GreaseService {
  /**
   * Application name.
   *
   * @public
   * @static
   * @readonly
   * @member {string} NAME
   */
  public static readonly NAME: string = ConfigService.NAME

  /**
   * Create a new grease runner service.
   *
   * @param {ConfigService} gc - Configuration service
   * @param {EventBus} events - Events bus
   * @param {CommandBus} operations - Operations bus
   * @param {QueryBus} queries - Query bus
   * @param {LoggerService} logger - Logger service
   */
  constructor(
    protected readonly gc: ConfigService,
    protected readonly events: EventBus,
    protected readonly operations: CommandBus,
    protected readonly queries: QueryBus,
    public readonly logger: LoggerService
  ) {}

  /**
   * Bump a package version.
   *
   * @see {@linkcode BumpEvent}
   * @see {@linkcode BumpOperationDTO}
   * @see {@linkcode PackageManifest}
   *
   * @public
   * @async
   * @fires BumpEvent
   *
   * @param {BumpOperationDTO} payload - Bump operation payload
   * @return {Promise<PackageManifest>} Package manifest
   */
  public async bump(payload: BumpOperationDTO): Promise<PackageManifest> {
    /**
     * Bump operation to execute.
     *
     * @const {BumpOperation} operation
     */
    const operation: BumpOperation = new BumpOperation(payload)

    /**
     * Updated package manifest.
     *
     * @const {PackageManifest} manifest
     */
    const manifest: PackageManifest = await this.operations.execute(operation)

    this.events.publish(new BumpEvent(manifest, operation))
    return manifest
  }

  /**
   * Generate a changelog from git metadata.
   *
   * @see {@linkcode ChangelogEvent}
   * @see {@linkcode ChangelogOperationDTO}
   * @see {@linkcode ChangelogStream}
   *
   * @public
   * @async
   * @fires ChangelogEvent
   *
   * @template T - Parsed commit type
   *
   * @param {ChangelogOperationDTO<T>?} [payload] - Changelog operation payload
   * @return {Promise<ChangelogStream<T>>} Changlog stream
   */
  public async changelog<T extends Commit = Commit>(
    payload?: ChangelogOperationDTO<T>
  ): Promise<ChangelogStream<T>> {
    /**
     * Changelog operation to execute.
     *
     * @const {ChangelogOperation<T>} operation
     */
    const operation: ChangelogOperation<T> = new ChangelogOperation(payload)

    /**
     * Changelog stream.
     *
     * @const {ChangelogStream<T>}
     */
    const stream: ChangelogStream<T> = await this.operations.execute(operation)

    this.events.publish(new ChangelogEvent<T>(stream, operation))
    return stream
  }

  /**
   * Load configuration options from a grease config file.
   *
   * @see {@linkcode GreaseConfig}
   * @see {@linkcode IGreaseConfig}
   *
   * @public
   * @async
   *
   * @template T - Parsed commit type
   *
   * @param {IGreaseConfig<T>?} [opts] - Configuration overrides
   * @return {Promise<EmptyObject | GreaseConfig<T>>} Configuration options
   */
  public async config<T extends Commit = Commit>(
    opts?: IGreaseConfig<T>
  ): Promise<EmptyObject | GreaseConfig<T>> {
    const { config } = new GlobalOptions(opts)
    return isString(config) ? this.gc.load(config, opts) : this.gc.search(opts)
  }

  /**
   * Get a version bump recommendation.
   *
   * @see {@linkcode BumpEvent}
   * @see {@linkcode BumpQuery}
   * @see {@linkcode RecommendedBump}
   *
   * @public
   * @async
   * @fires BumpEvent
   *
   * @param {Partial<BumpQuery>?} [params] - Version bump query parameters
   * @return {Promise<RecommendedBump>} Recommended version bump
   */
  public async recommend(
    params?: Partial<BumpQuery>
  ): Promise<RecommendedBump> {
    /**
     * Version bump query to execute.
     *
     * @const {BumpQuery} query
     */
    const query: BumpQuery = new BumpQuery(params)

    /**
     * Version bump recommendation.
     *
     * @const {RecommendedBump} recommendation
     */
    const recommendation: RecommendedBump = await this.queries.execute(query)

    this.events.publish(new BumpEvent(recommendation, query))
    return recommendation
  }

  /**
   * Create a new release tag, or overwrite an existing one.
   *
   * @see {@linkcode TagOperationDTO}
   * @see {@linkcode TagOperation}
   *
   * @public
   * @async
   *
   * @param {TagOperationDTO} payload - Tag operation payload
   * @return {Promise<TagOperation>} Executed tag operation
   */
  public async tag(payload: TagOperationDTO): Promise<TagOperation> {
    return this.operations.execute(new TagOperation(payload))
  }

  /**
   * Get git tags.
   *
   * @see {@linkcode TagQuery}
   *
   * @public
   * @async
   *
   * @param {Partial<TagQuery>?} [params] - Tag query parameters
   * @return {Promise<string[]>} Git tags array
   */
  public async tags(params?: Partial<TagQuery>): Promise<string[]> {
    return this.queries.execute(new TagQuery(params))
  }
}

export default GreaseService
