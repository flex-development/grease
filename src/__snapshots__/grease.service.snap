// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:GreaseService > #bump > should return bump version 1`] = `
{
  "build": [],
  "major": 3,
  "minor": 0,
  "patch": 0,
  "prerelease": [
    "rc",
    1,
  ],
  "version": "3.0.0-rc.1",
}
`;

exports[`unit:GreaseService > #recommend > should return recommended version bump 1`] = `
{
  "breaks": 1,
  "bump": "premajor",
  "commits": 72,
  "features": 16,
  "unstable": true,
}
`;

exports[`unit:GreaseService > #tag > should return executed tag operation 1`] = `
TagOperation {
  "color": true,
  "config": true,
  "cwd": "$PWD",
  "force": false,
  "level": "warn",
  "message": "",
  "object": "HEAD",
  "push": false,
  "remote": "origin",
  "sign": true,
  "tag": "grease@3.0.0-alpha.1",
  "tagprefix": "grease@",
  "unstable": true,
  "verify": true,
}
`;

exports[`unit:GreaseService > #tags > should return tags array 1`] = `
[
  "grease@2.0.0",
  "grease@1.1.0",
  "grease@1.0.2",
  "grease@1.0.1",
  "grease@1.0.0",
]
`;
